### PR TITLE
Add autocomplete support for Module and struct.dataclass.

### PR DESCRIPTION
**Pull request by jheek**
image

### CODE REVIEW

File: flax/linen/module.py [hunk 1]
```diff
@@ -51,6 +52,20 @@
 _CallableT = TypeVar('_CallableT', bound=Callable)
 
 
+_T = TypeVar("_T")
+
+def __dataclass_transform__(
```
Comment (andsteing) [inline flax/linen/module.py:57]: You already import this?

File: flax/struct.py [hunk 2]
```diff
@@ -39,7 +39,24 @@
 import jax
 
 
-def dataclass(clz: type):
+
+# This decorator is interpreted by static analysis tools as a hint
+# that a decorator or metaclass causes dataclass-like behavior.
+_T = TypeVar("_T")
+def __dataclass_transform__(
```
Comment (andsteing) [inline flax/struct.py:46]: Should we really add dunders to this?

I don't know that from other libraries, maybe you found some guidance on when to use them?

**Review by andsteing — APPROVED** [review]
It's really great that this works, but I don't understand from your comments *why* it's working. If you could extend a bit in the code or the PR that would be interesting...

**Comment by avital** [comment]
Yes adding a comment in the code (not the PR) is ideal so that future
readers immediately can understand

File: flax/linen/module.py (hunk 1 above)
Comment (jheek) [reply flax/linen/module.py]: fixed

File: flax/struct.py (hunk 2 above)
Comment (jheek) [reply flax/struct.py:48]: Yes, I've added a link to the proposal that explains this magic

File: tests/linen/module_test.py [hunk 3]
```diff
@@ -1371,15 +1370,18 @@ class Foo(NamedTuple):
...
-    T = TypeVar('T')
-    class MyComponent(nn.Module, Generic[T]):
-      pass
-    class MyModule(nn.Module):
-      submodule: MyComponent[jnp.ndarray]
-    class MyComponent2(Generic[T], nn.Module):
-      pass
-    class MyModule2(nn.Module):
-      submodule: MyComponent2[jnp.ndarray]
+    if sys.version_info.major == 3 and sys.version_info.minor >= 7:
```
Comment (andsteing) [inline tests/linen/module_test.py:1373]: So this bug only appeared with the added `__dataclass_transform__`?

**Review by andsteing — APPROVED** [review]

File: tests/linen/module_test.py (hunk 3 above)
Comment (jheek) [reply tests/linen/module_test.py:1373]: Yes because the __dataclass_transform__ only works on metaclasses so we have an empty metaclass. In the exceptional edge case where somebody wants do multiple inheritance with Generic on a Module they could probably use `six`

**Comment by cgarciae** [comment]
This is going to be a huge quality-of-life improvement for Flax!

File: tests/linen/module_test.py (hunk 3 above)
Comment (jheek) [reply tests/linen/module_test.py]: Refactored the code to only add the metaclass during type checking. This avoids the problem with metaclass inheritance
